src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -148,11 +160,108 @@ protected override void OnStopped()
             // the other is external thing disconnecting remote host from us (ex, user killing OOP process).
             // the Disconnected event we subscribe is to detect #2 case. and this method is for #1 case. so when we are willingly disconnecting
             // we don't need the event, otherwise, Disconnected event will be called twice.
+            UnregisterGlobalOperationNotifications();
             _rpc.Disconnected -= OnRpcDisconnected;
             _rpc.Dispose();
             _remotableDataRpc.Dispose();
         }
 
+        private void RegisterGlobalOperationNotifications()
+        {
+            var globalOperationService = this.Workspace.Services.GetService<IGlobalOperationNotificationService>();
+            if (globalOperationService != null)
+            {
+                globalOperationService.Started += OnGlobalOperationStarted;
+                globalOperationService.Stopped += OnGlobalOperationStopped;
+            }
+        }
+
+        private void UnregisterGlobalOperationNotifications()
+        {
+            var globalOperationService = this.Workspace.Services.GetService<IGlobalOperationNotificationService>();
+            if (globalOperationService != null)
+            {
+                globalOperationService.Started -= OnGlobalOperationStarted;
+                globalOperationService.Stopped -= OnGlobalOperationStopped;
+            }
+
+            lock (_globalNotificationsGate)
+            {
+                // Unilaterally transition us to the finished state.  Once we're finished
+                // we cannot start or stop anymore.
+                _globalNotificationsTask = _globalNotificationsTask.ContinueWith(
+                    _ => GlobalNotificationState.Finished, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
+            }
+        }
+
+        private void OnGlobalOperationStarted(object sender, EventArgs e)
+        {
+            lock (_globalNotificationsGate)
+            {
+                _globalNotificationsTask = _globalNotificationsTask.ContinueWith(
+                    continuation, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default).Unwrap();
+            }
+
+            async Task<GlobalNotificationState> continuation(Task<GlobalNotificationState> previousTask)
+            {
+                // Can only transition from NotStarted->Started.  If we hear about
+                // anything else, do nothing.
+                if (previousTask.Result != GlobalNotificationState.NotStarted)
+                {
+                    return previousTask.Result;
+                }
+
+                using (var connection = await this.TryCreateConnectionAsync(WellKnownRemoteHostServices.RemoteHostService, CancellationToken.None).ConfigureAwait(false))
```

Comment:
oh.  so we always have the RemoteHostService as the receiver of a client's rpc calls?  it's not some other object htat is the receiver?